File: backend-api/app/services/user_service.py
"""
用户服务
"""
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from typing import List, Optional, Dict, Any
import uuid
from app.models.user import User, Role, UserStatus
from app.services.auth_service import AuthService
from app.core.exceptions import NotFoundException, ConflictException, ValidationException
from app.core.logging import logger
from app.utils.password_validator import validate_password_strength


class UserService:
    """用户服务类"""
    
    @staticmethod
    async def create_user(
        db: AsyncSession,
        username: str,
        password: str,
        email: str,
        full_name: str,
        department: Optional[str] = None,
        position: Optional[str] = None,
        phone: Optional[str] = None
    ) -> User:
        """
        创建用户
        
        Args:
            db: 数据库会话
            username: 用户名
            password: 密码
            email: 邮箱
            full_name: 真实姓名
            department: 部门
            position: 职位
            phone: 电话
            
        Returns:
            User: 用户对象
            
        Raises:
            ConflictException: 用户名或邮箱已存在
            ValidationException: 数据验证失败
        """
        # 验证密码强度
        validate_password_strength(password, username)
        
        # 检查用户名是否已存在
        result = await db.execute(
            select(User).where(User.username == username)
        )
        if result.scalar_one_or_none():
            raise ConflictException(
                message=f"用户名已存在: {username}"
            )
        
        # 检查邮箱是否已存在
        result = await db.execute(
            select(User).where(User.email == email)
        )
        if result.scalar_one_or_none():
            raise ConflictException(
                message=f"邮箱已存在: {email}"
            )
        
        # 创建用户
        user = User(
            id=str(uuid.uuid4()),
            username=username,
            passwordHash=AuthService.hash_password(password),
            email=email,
            fullName=full_name,
            department=department,
            position=position,
            phone=phone,
            status=UserStatus.ACTIVE
        )
        
        db.add(user)
        await db.commit()
        await db.refresh(user)
        
        logger.info(f"User created: {username}")
        
        return user
    
    @staticmethod
    async def get_user_by_id(
        db: AsyncSession,
        user_id: str
    ) -> Optional[User]:
        """
        根据ID获取用户
        
        Args:
            db: 数据库会话
            user_id: 用户ID
            
        Returns:
            Optional[User]: 用户对象或 None
        """
        result = await db.execute(
            select(User).where(User.id == user_id)
        )
        return result.scalar_one_or_none()
    
    @staticmethod
    async def get_users(
        db: AsyncSession,
        skip: int = 0,
        limit: int = 100,
        status: Optional[UserStatus] = None,
        department: Optional[str] = None
    ) -> tuple[List[User], int]:
        """
        查询用户列表
        
        Args:
            db: 数据库会话
            skip: 跳过记录数
            limit: 返回记录数
            status: 用户状态（可选）
            department: 部门（可选）
            
        Returns:
            tuple[List[User], int]: 用户列表和总数
        """
        query = select(User)
        count_query = select(func.count()).select_from(User)
        
        if status:
            query = query.where(User.status == status)
            count_query = count_query.where(User.status == status)
        
        if department:
            query = query.where(User.department == department)
            count_query = count_query.where(User.department == department)
        
        # 获取总数
        result = await db.execute(count_query)
        total = result.scalar()
        
        # 获取用户列表
        query = query.offset(skip).limit(limit)
        result = await db.execute(query)
        users = list(result.scalars().all())
        
        return users, total
    
    @staticmethod
    async def update_user(
        db: AsyncSession,
        user_id: str,
        email: Optional[str] = None,
        full_name: Optional[str] = None,
        department: Optional[str] = None,
        position: Optional[str] = None,
        phone: Optional[str] = None,
        status: Optional[UserStatus] = None
    ) -> User:
        """
        更新用户
        
        Args:
            db: 数据库会话
            user_id: 用户ID
            email: 邮箱（可选）
            full_name: 真实姓名（可选）
            department: 部门（可选）
            position: 职位（可选）
            phone: 电话（可选）
            status: 状态（可选）
            
        Returns:
            User: 更新后的用户对象
            
        Raises:
            NotFoundException: 用户不存在
            ConflictException: 邮箱已存在
        """
        user = await UserService.get_user_by_id(db, user_id)
        
        if not user:
            raise NotFoundException(
                message="用户不存在",
                error_code="USER_NOT_FOUND"
            )
        
        # 检查邮箱是否重复
        if email and email != user.email:
            result = await db.execute(
                select(User).where(User.email == email)
            )
            if result.scalar_one_or_none():
                raise ConflictException(
                    message=f"邮箱已存在: {email}"
                )
            user.email = email
        
        if full_name is not None:
            user.fullName = full_name
        
        if department is not None:
            user.department = department
        
        if position is not None:
            user.position = position
        
        if phone is not None:
            user.phone = phone
        
        if status is not None:
            user.status = status
        
        await db.commit()
        await db.refresh(user)
        
        logger.info(f"User updated: {user_id}")
        
        return user
    
    @staticmethod
    async def delete_user(
        db: AsyncSession,
        user_id: str
    ) -> None:
        """
        删除用户
        
        Args:
            db: 数据库会话
            user_id: 用户ID
            
        Raises:
            NotFoundException: 用户不存在
        """
        user = await UserService.get_user_by_id(db, user_id)
        
        if not user:
            raise NotFoundException(
                message="用户不存在",
                error_code="USER_NOT_FOUND"
            )
        
        await db.delete(user)
        await db.commit()
        
        logger.info(f"User deleted: {user_id}")
    
    @staticmethod
    async def assign_roles(
        db: AsyncSession,
        user_id: str,
        role_ids: List[str]
    ) -> User:
        """
        为用户分配角色
        
        Args:
            db: 数据库会话
            user_id: 用户ID
            role_ids: 角色ID列表
            
        Returns:
            User: 更新后的用户对象
            
        Raises:
            NotFoundException: 用户或角色不存在
        """
        user = await UserService.get_user_by_id(db, user_id)
        
        if not user:
            raise NotFoundException(
                message="用户不存在",
                error_code="USER_NOT_FOUND"
            )
        
        # 获取角色对象
        roles = []
        for role_id in role_ids:
            result = await db.execute(
                select(Role).where(Role.id == role_id)
            )
            role = result.scalar_one_or_none()
            
            if not role:
                raise NotFoundException(
                    message=f"角色不存在: {role_id}",
                    error_code="ROLE_NOT_FOUND"
                )
            
            roles.append(role)
        
        # 分配角色
        user.roles = roles
        
        await db.commit()
        await db.refresh(user)
        
        logger.info(f"Roles assigned to user {user_id}: {len(roles)} roles")
        
        return user
    
    @staticmethod
    async def get_user_roles(
        db: AsyncSession,
        user_id: str
    ) -> List[Role]:
        """
        获取用户的角色列表
        
        Args:
            db: 数据库会话
            user_id: 用户ID
            
        Returns:
            List[Role]: 角色列表
            
        Raises:
            NotFoundException: 用户不存在
        """
        user = await UserService.get_user_by_id(db, user_id)
        
        if not user:
            raise NotFoundException(
                message="用户不存在",
                error_code="USER_NOT_FOUND"
            )
        
        # 加载用户的角色
        await db.refresh(user, ['roles'])
        
        return user.roles
    
    @staticmethod
    async def change_password(
        db: AsyncSession,
        user_id: str,
        old_password: str,
        new_password: str
    ) -> None:
        """
        修改密码
        
        Args:
            db: 数据库会话
            user_id: 用户ID
            old_password: 旧密码
            new_password: 新密码
            
        Raises:
            NotFoundException: 用户不存在
            ValidationException: 旧密码错误或新密码不符合要求
        """
        user = await UserService.get_user_by_id(db, user_id)
        
        if not user:
            raise NotFoundException(
                message="用户不存在",
                error_code="USER_NOT_FOUND"
            )
        
        # 验证旧密码
        if not AuthService.verify_password(old_password, user.passwordHash):
            raise ValidationException(
                message="旧密码错误"
            )
        
        # 验证新密码强度
        validate_password_strength(new_password, user.username)
        
        # 更新密码
        user.passwordHash = AuthService.hash_password(new_password)
        
        await db.commit()
        
        logger.info(f"Password changed for user: {user_id}")

File: backend-api-deprecated/src/services/userService.ts
import { PrismaClient, UserStatus } from '@prisma/client'
import { AuthService } from './authService'
import {
  CreateUserDto,
  UpdateUserDto,
  UserQuery,
  ResetPasswordDto,
  UserResponse,
  PaginatedUserResult
} from '../types/user'
import { logger } from '../config/logger'

const prisma = new PrismaClient()

export class UserService {
  /**
   * 创建用户
   * @param data 用户创建数据
   * @param createdBy 创建者ID
   * @returns 创建的用户信息
   */
  async createUser(data: CreateUserDto, createdBy: string): Promise<UserResponse> {
    try {
      // 检查用户名是否已存在
      const existingUser = await prisma.user.findUnique({
        where: { username: data.username }
      })

      if (existingUser) {
        throw new Error('用户名已存在')
      }

      // 检查邮箱是否已存在
      const existingEmail = await prisma.user.findUnique({
        where: { email: data.email }
      })

      if (existingEmail) {
        throw new Error('邮箱已被使用')
      }

      // 验证角色是否存在
      if (data.roleIds && data.roleIds.length > 0) {
        const roles = await prisma.role.findMany({
          where: { id: { in: data.roleIds } }
        })

        if (roles.length !== data.roleIds.length) {
          throw new Error('部分角色不存在')
        }
      }

      // 哈希密码
      const passwordHash = await AuthService.hashPassword(data.password)

      // 创建用户（使用事务）
      const user = await prisma.$transaction(async (tx) => {
        // 创建用户
        const newUser = await tx.user.create({
          data: {
            username: data.username,
            passwordHash,
            email: data.email,
            fullName: data.fullName,
            department: data.department,
            position: data.position,
            phone: data.phone,
            status: UserStatus.ACTIVE
          }
        })

        // 分配角色
        if (data.roleIds && data.roleIds.length > 0) {
          await tx.userRole.createMany({
            data: data.roleIds.map(roleId => ({
              userId: newUser.id,
              roleId
            }))
          })
        }

        return newUser
      })

      logger.info('User created', { userId: user.id, username: user.username, createdBy })

      // 查询完整的用户信息（包含角色）
      return await this.getUserById(user.id)
    } catch (error) {
      logger.error('Create user failed', { error, data: { ...data, password: '[REDACTED]' } })
      throw error
    }
  }

  /**
   * 更新用户信息
   * @param userId 用户ID
   * @param data 更新数据
   * @param updatedBy 更新者ID
   * @returns 更新后的用户信息
   */
  async updateUser(userId: string, data: UpdateUserDto, updatedBy: string): Promise<UserResponse> {
    try {
      // 检查用户是否存在
      const existingUser = await prisma.user.findUnique({
        where: { id: userId }
      })

      if (!existingUser) {
        throw new Error('用户不存在')
      }

      // 如果更新邮箱，检查是否已被使用
      if (data.email && data.email !== existingUser.email) {
        const emailInUse = await prisma.user.findUnique({
          where: { email: data.email }
        })

        if (emailInUse) {
          throw new Error('邮箱已被使用')
        }
      }

      // 验证角色是否存在
      if (data.roleIds && data.roleIds.length > 0) {
        const roles = await prisma.role.findMany({
          where: { id: { in: data.roleIds } }
        })

        if (roles.length !== data.roleIds.length) {
          throw new Error('部分角色不存在')
        }
      }

      // 更新用户（使用事务）
      const user = await prisma.$transaction(async (tx) => {
        // 更新用户基本信息
        const updatedUser = await tx.user.update({
          where: { id: userId },
          data: {
            email: data.email,
            fullName: data.fullName,
            department: data.department,
            position: data.position,
            phone: data.phone,
            status: data.status
          }
        })

        // 更新角色（如果提供了 roleIds）
        if (data.roleIds !== undefined) {
          // 删除现有角色关联
          await tx.userRole.deleteMany({
            where: { userId }
          })

          // 创建新的角色关联
          if (data.roleIds.length > 0) {
            await tx.userRole.createMany({
              data: data.roleIds.map(roleId => ({
                userId,
                roleId
              }))
            })
          }
        }

        return updatedUser
      })

      logger.info('User updated', { userId, updatedBy })

      // 查询完整的用户信息（包含角色）
      return await this.getUserById(user.id)
    } catch (error) {
      logger.error('Update user failed', { error, userId, data })
      throw error
    }
  }

  /**
   * 获取用户详情
   * @param userId 用户ID
   * @returns 用户信息
   */
  async getUserById(userId: string): Promise<UserResponse> {
    const user = await prisma.user.findUnique({
      where: { id: userId },
      include: {
        roles: {
          include: {
            role: true
          }
        }
      }
    })

    if (!user) {
      throw new Error('用户不存在')
    }

    return {
      id: user.id,
      username: user.username,
      email: user.email,
      fullName: user.fullName,
      department: user.department || undefined,
      position: user.position || undefined,
      phone: user.phone || undefined,
      status: user.status,
      roles: user.roles.map(ur => ({
        id: ur.role.id,
        name: ur.role.name,
        description: ur.role.description || undefined
      })),
      createdAt: user.createdAt,
      updatedAt: user.updatedAt,
      lastLoginAt: user.lastLoginAt || undefined
    }
  }

  /**
   * 查询用户列表
   * @param query 查询参数
   * @returns 分页的用户列表
   */
  async listUsers(query: UserQuery): Promise<PaginatedUserResult> {
    const {
      page = 1,
      pageSize = 20,
      username,
      email,
      fullName,
      department,
      status,
      roleId
    } = query

    // 构建查询条件
    const where: any = {}

    if (username) {
      where.username = { contains: username, mode: 'insensitive' }
    }

    if (email) {
      where.email = { contains: email, mode: 'insensitive' }
    }

    if (fullName) {
      where.fullName = { contains: fullName, mode: 'insensitive' }
    }

    if (department) {
      where.department = { contains: department, mode: 'insensitive' }
    }

    if (status) {
      where.status = status
    }

    if (roleId) {
      where.roles = {
        some: {
          roleId
        }
      }
    }

    // 计算分页参数
    const skip = (page - 1) * pageSize
    const take = pageSize

    // 查询用户列表和总数
    const [users, total] = await Promise.all([
      prisma.user.findMany({
        where,
        skip,
        take,
        include: {
          roles: {
            include: {
              role: true
            }
          }
        },
        orderBy: {
          createdAt: 'desc'
        }
      }),
      prisma.user.count({ where })
    ])

    // 转换为响应格式
    const items: UserResponse[] = users.map(user => ({
      id: user.id,
      username: user.username,
      email: user.email,
      fullName: user.fullName,
      department: user.department || undefined,
      position: user.position || undefined,
      phone: user.phone || undefined,
      status: user.status,
      roles: user.roles.map(ur => ({
        id: ur.role.id,
        name: ur.role.name,
        description: ur.role.description || undefined
      })),
      createdAt: user.createdAt,
      updatedAt: user.updatedAt,
      lastLoginAt: user.lastLoginAt || undefined
    }))

    return {
      items,
      total,
      page,
      pageSize,
      totalPages: Math.ceil(total / pageSize)
    }
  }

  /**
   * 更新用户状态
   * @param userId 用户ID
   * @param status 新状态
   * @param updatedBy 更新者ID
   * @returns 更新后的用户信息
   */
  async updateUserStatus(userId: string, status: UserStatus, updatedBy: string): Promise<UserResponse> {
    try {
      // 检查用户是否存在
      const existingUser = await prisma.user.findUnique({
        where: { id: userId }
      })

      if (!existingUser) {
        throw new Error('用户不存在')
      }

      // 更新状态
      await prisma.user.update({
        where: { id: userId },
        data: { status }
      })

      logger.info('User status updated', { userId, status, updatedBy })

      // 返回更新后的用户信息
      return await this.getUserById(userId)
    } catch (error) {
      logger.error('Update user status failed', { error, userId, status })
      throw error
    }
  }

  /**
   * 重置用户密码（管理员操作）
   * @param userId 用户ID
   * @param data 密码重置数据
   * @param resetBy 重置者ID
   */
  async resetPassword(userId: string, data: ResetPasswordDto, resetBy: string): Promise<void> {
    try {
      // 检查用户是否存在
      const user = await prisma.user.findUnique({
        where: { id: userId }
      })

      if (!user) {
        throw new Error('用户不存在')
      }

      // 哈希新密码
      const passwordHash = await AuthService.hashPassword(data.newPassword)

      // 更新密码
      await prisma.user.update({
        where: { id: userId },
        data: { passwordHash }
      })

      logger.info('Password reset', { userId, resetBy })
    } catch (error) {
      logger.error('Reset password failed', { error, userId })
      throw error
    }
  }

  /**
   * 删除用户（软删除 - 设置为 INACTIVE）
   * @param userId 用户ID
   * @param deletedBy 删除者ID
   */
  async deleteUser(userId: string, deletedBy: string): Promise<void> {
    try {
      // 检查用户是否存在
      const user = await prisma.user.findUnique({
        where: { id: userId }
      })

      if (!user) {
        throw new Error('用户不存在')
      }

      // 软删除：设置状态为 INACTIVE
      await prisma.user.update({
        where: { id: userId },
        data: { status: UserStatus.INACTIVE }
      })

      logger.info('User deleted (soft)', { userId, deletedBy })
    } catch (error) {
      logger.error('Delete user failed', { error, userId })
      throw error
    }
  }
}

export const userService = new UserService()
